src/debugger/SteppingManager.js

```diff
@@ -11,50 +11,88 @@
     this._channel = channel;
+    this._realm = realm;
+    this._steppers = [];
+    this._keepOldSteppers = false;
+    if (keepOldSteppers) this._keepOldSteppers = true;
   }
   _channel: DebugChannel;
-  _stepInto: void | StepIntoStepper;
+  _realm: Realm;
+  _keepOldSteppers: boolean;
+  _steppers: Array<Stepper>;
 
   processStepCommand(kind: "in" | "over" | "out", currentNode: BabelNode) {
     if (kind === "in") {
       this._processStepIn(currentNode);
+    } else if (kind === "over") {
+      this._processStepOver(currentNode);
     }
     // TODO: implement stepOver and stepOut
   }
 
   _processStepIn(ast: BabelNode) {
     invariant(this._stepInto === undefined);
     invariant(ast.loc && ast.loc.source);
-    this._stepInto = new StepIntoStepper(ast.loc.source, ast.loc.start.line, ast.loc.start.column);
+    if (!this._keepOldSteppers) {
+      this._steppers = [];
+    }
+    this._steppers.push(new StepIntoStepper(ast.loc.source, ast.loc.start.line, ast.loc.start.column));
+  }
+
+  _processStepOver(ast: BabelNode) {
+    invariant(ast.loc && ast.loc.source);
+    if (!this._keepOldSteppers) {
+      this._steppers = [];
+    }
+    this._steppers.push(
+      new StepOverStepper(ast.loc.source, ast.loc.start.line, ast.loc.start.column, this._realm.contextStack.length)
+    );
   }
 
-  isStepComplete(ast: BabelNode): boolean {
-    if (ast.loc && ast.loc.source) {
-      if (this._stepInto && this._stepInto.isComplete(ast)) {
-        this._stepInto = undefined;
-        this._channel.sendStoppedResponse("Step Into", ast.loc.source, ast.loc.start.line, ast.loc.start.column);
-        return true;
+  isStepComplete(ast: BabelNode): void | SteppingType {
+    let completedStepperType = this._anyStepperComplete(ast);
+    if (completedStepperType) {
+      invariant(ast.loc && ast.loc.source);
+      this._channel.sendStoppedResponse(completedStepperType, ast.loc.source, ast.loc.start.line, ast.loc.start.column);
+      return completedStepperType;
+    }
+    return undefined;
+  }
+
+  _anyStepperComplete(ast: BabelNode): void | SteppingType {
+    let i = 0;
+    let completedStepperType;
+    while (i < this._steppers.length) {
+      let stepper = this._steppers[i];
+      if (stepper.isComplete(ast, this._realm)) {
+        if (stepper instanceof StepIntoStepper) completedStepperType = "Step Into";
+        if (stepper instanceof StepOverStepper) completedStepperType = "Step Over";
+        this._steppers.splice(i, 1);
+      } else {
+        i++;
       }
     }
-    return false;
+    return completedStepperType;
   }
 
   onDebuggeeStop(ast: BabelNode, reason: StoppedReason) {
-    if (reason !== "Step Into") {
-      // stopped for another reason, e.g. breakpoint
-      if (this._stepInto !== undefined) {
-        // we're in the middle of a step into, but debuggee has stopped for another reason here first, so cancel this step into
-        this._stepInto = undefined;
+    if (reason === "Breakpoint") {
```

Comment:
I do not like this design of arbitration on stopping logic. It hard-coded the assumption that breakpoint manager is stopped before steppingmanager which may change in future. For example, when we added exception manager to deal with exception, who is responsible for deciding stop? 

There should be a dedicated StopEventManager that gathers information from all managers and make the central decision.

Reply:
I don't think it makes this assumption. The onDebuggeeStopped in each manager is called from the _onDebuggeeStopped method of DebugServer. This happens immediately before Prepack goes into a waiting state for any reason, so it does not make any assumptions on which order the managers were checked. If we would like to have some aggregation of the stopping, we can have each of the manager's onDebuggeeStopped methods return a result and the _onDebuggeeStopped in the DebugServer can be changed into a StopEventManager that joins these results.

Reply:
The fact that you have to explicitly check "reason === "Breakpoint"" here indicates you have some hard-code assumption in this design.